### PR TITLE
Refactored to use getopts in some places.

### LINKED ISSUE #42: Update cromshell to use getopts
We live in the future, so our arg parsing should at least live in the present.

### LINKED ISSUE #43: Sometimes the mascot is shown twice.
Turtles are nice
but not when they're twice.
One is enough,
Two is too rough.

### PR DESCRIPTION
**Pull request by jonn-smith**
Could not remove all other arg parsing due to positional args.

Fixed #43 - Twyseturtles was due to the `notify` logic.
Fixed #42 

@lbergelson - I tested it reasonably well locally, but it could definitely use more testing.  In particular, `submit`, `notify`, and `list` (but really a little of everything).

### CODE REVIEW

File: cromshell [hunk 1]
```diff
@@ -294,15 +371,24 @@ function populateWorkflowIdAndServerUrl()
...
-      assertHavePreviouslySubmittedAJob
-      WORKFLOW_ID=$( tail -n1 ${CROMWELL_SUBMISSIONS_FILE} | awk '{print $3}' )
+      populateLastWorkflowId
       WORKFLOW_SERVER_URL=$(tail -n1 $CROMWELL_SUBMISSIONS_FILE | awk '{print $2}' )
     fi
   fi
+
+  # Validate our workflow ID:
+  if [[ $( matchWorkflowId ${WORKFLOW_ID} ) -ne 0 ]] ; then
+    if [ -n "$userSpecifiedId" ] ; then
```
Comment (lbergelson) [inline cromshell]: seems like a good check to add

File: cromshell [hunk 2]
```diff
@@ -432,19 +518,29 @@ function abort()
...
+        ;;
+      u)
+        doUpdate=true
+        ;;
+      *)
+        invalidSubCommand list flag ${OPTARG}
+        ;; 
+    esac 
+  done
+  shift $((OPTIND-1))
```
Comment (lbergelson) [inline cromshell]: I don't really know how getopts works.  What's this shift doing after the opt parsing loop? Should it go within the opt parsing loop to loop over all the options?

Comment (lbergelson) [reply cromshell]: Also, where is `opt` defined?

File: cromshell [hunk 3]
```diff
@@ -608,20 +704,86 @@ function fetch-all()
...
+    error "Error: You must specify an email address to notify."
+    error ""
+    simpleUsage notify
+    echo "Try \`$SCRIPTNAME -h' for more information."
+    exit 5
+  fi
 
+  # Is the next argument our workflow ID?
+  # Workflow IDs are standard UUIDs or negative numbers:
+  if [[ $(matchWorkflowId ${1}) -eq 0 ]] || [[ ${1} == -* ]] ; then
```
Comment (lbergelson) [inline cromshell]: does `[[ ${1} == -* ]]` read as: " the value of $1 is equal to a `-` followed by any characters" ?  If so should it be a match of `-[:digit:]+` instead?

File: cromshell [hunk 4]
```diff
@@ -608,20 +704,86 @@ function fetch-all()
...
+    shift
+  fi
+
+  # Make sure the user gave us a good email address:
+  echo "$1" | grep -q  '@' 
+  r=$?
+  if [ $r -ne 0 ] ; then
+    error "Error: invalid email address: $1"
+    error ""
+    simpleUsage notify
```
Comment (lbergelson) [inline cromshell]: Maybe pull this assert out as a function `assertValidEmailAddress`?

File: cromshell [hunk 5]
```diff
@@ -649,127 +811,66 @@ function runSubCommandOnWorkflowId()
...
-        ;;
-      submit|status|logs|execution-status-count|metadata|slim-metadata|timing|abort|notify|list|fetch-all|fetch-logs|list-outputs) 
-        # Get our sub-command:
-        SUB_COMMAND=$1
-        shift
-        # We'll manually handle the rest of the options:
-        break
+  # Get top-level arguments to script itself:
+  while getopts "h" opt ; do
+    case ${opt} in
```
Comment (lbergelson) [inline cromshell]: This help checking seems redundant / strange considering the more advanced help check you added elsewhere.

File: cromshell [hunk 6]
```diff
@@ -649,127 +811,66 @@ function runSubCommandOnWorkflowId()
...
+  shift $((OPTIND-1)) 
 
-  ################################################################################
+  # Get our sub-command:
+  SUB_COMMAND=${1}
+  shift
 
-  # Make sure we have all the required commands:
+  # Validate our sub-command:
+  case ${SUB_COMMAND} in
```
Comment (lbergelson) [inline cromshell]: additional help check, shouldn't that all be consolidated where you ran the advanced help check?

File: cromshell [hunk 7]
```diff
@@ -649,127 +811,66 @@ function runSubCommandOnWorkflowId()
...
-    shift
-  done
-
-  # Exit the script with the return value of the last run sub-command:
-  exit $r
-
+  # Handle specific sub-command args and and call our sub-command:
+  echo "Sub-Command: ${SUB_COMMAND}"
+  case ${SUB_COMMAND} in 
+    submit|list|notify)
```
Comment (lbergelson) [inline cromshell]: this is a huge improvement over the previous status quo

File: cromshell [hunk 8]
```diff
@@ -182,15 +191,25 @@ function makeTemp()
...
 {
   cleanTempVars
 }
+# Make sure we clean up on exit:
+trap at_exit EXIT 
+
+# Print the name and value of a simple variable:
+function printVar()
+{
+  error "$1 = ${!1}"
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply cromshell]: :D

File: cromshell [hunk 9]
```diff
@@ -243,6 +262,64 @@ function checkPrerequisites
...
+      shift
+    done
+  fi
+}
+
+function matchWorkflowId() 
+{
+  local rv=1
+  
+  if [[ "${1}" =~ ^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$ ]] ; then
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply cromshell]: I specifically want to validate only on valid `UUID`s, so I need to exclude characters `g-z`.

File: cromshell (hunk 1 above)
Comment (jonn-smith) [reply cromshell]: Thanks.  I thought so too :P

File: cromshell (hunk 2 above)
Comment (jonn-smith) [reply cromshell]: It's just a more sane option parser.  The only problem is that while it works well for single character flags, it doesn't work well for positional args or long args (e.g. `--help`), so it ended up being a little less useful for cromshell as it stands now.

That `shift` is shifting the arguments by the number of args that have been parsed by `getopt`.

`opt` is defined at the top of the while loop (line 526): `while getopts "cu" `**opt**` ; do`

Comment (lbergelson) [reply cromshell]: Oh, it's defined by getopts... that makes sense... Does getopt default to using `$@` as it's input?

File: cromshell (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonn-smith) [reply cromshell]: xdigit works!

File: cromshell (hunk 2 above)
Comment (jonn-smith) [reply cromshell]: That's my understanding, which is how it can be used in functions (providing you mark `OPTIND` as local).

**Pull request merged**